[Improvement-17157][Master] Support setting max.concurrent.workflow.instances

tusaryan opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler, we are happy that you want to help us improve DolphinScheduler! -->

## Purpose of the pull request

<!--(For example: This pull request adds checkstyle plugin).-->
This Pull Request implements a new configuration option `max.concurrent.workflow.instances` in the master server, allowing administrators to set a limit on the number of concurrently running workflow instances on a single master. This addresses the issue of potential overload when multiple workflows failover to a single master, as described in #17157.

close #17157

## Brief change log


-   **Feature:** Implemented a new configuration property, `master.server-load-protection.max-concurrent-workflow-instances`, to limit concurrent workflows. The master is marked as busy if the number of running workflows is greater than or equal to this value.

-   **Enhancement:** Refactored `MasterServerLoadProtection` to incorporate system resource thresholds based on community feedback. The following configuration properties are now utilized to provide a more robust server protection mechanism:
    -   `maxSystemCpuUsagePercentageThresholds`
    -   `maxJvmCpuUsagePercentageThresholds`
    -   `maxSystemMemoryUsagePercentageThresholds`
    -   `maxDiskUsagePercentageThresholds`

-   **Bug Fix:** Resolved a `NullPointerException` that occurred during MasterServer startup by correctly using `@Autowired` to inject the `MasterServerLoadProtection` bean into `MasterConfig`.

-   **Refactor:**
    -   Created `MasterServerLoadProtectionConfig` to manage the bean's lifecycle and properties via Spring.
    -   Replaced the use of `WorkflowCacheRepository` with the `IWorkflowRepository` interface for better abstraction and testability.

-   **Documentation:** Updated the `application.yml` file and the `configuration.md` contributor documentation to include all new load protection options with clear descriptions. Example:
    ```yaml
    master:
      server-load-protection:
        # Master max concurrent workflow instances. When the count reaches or exceeds this value, the master is marked busy.
        max-concurrent-workflow-instances: 100  #Set your desired limit 
    ```
Replace 100 with any integer value you want as the limit.

## Verify this pull request

This change is covered by updated unit tests in `MasterServerLoadProtectionTest.java` to include comprehensive test cases verifying the new functionality for both the workflow instance limit and the system resource thresholds:

-   **Workflow Limit Test:** A dedicated test (`isOverloadWithMaxConcurrentWorkflowInstances`) verifies the new limit:
    -   When workflow count (e.g., 5) is **less than** the limit (e.g., 10), the master is **not** overloaded.
    -   When workflow count (e.g., 5) is **equal to or greater than** the limit (e.g., 5 or 3), the master **is** overloaded.
-   **System Metrics Test:** Tests have been updated to confirm that the master is correctly marked as overloaded when any of the system resource thresholds (CPU, memory, disk) are exceeded.


fix issue: [Improvement][Master] Support set max.concurrent.workflow.instances in master #17157


## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contains incompatible change, you should also add it to `docs/docs/en/guide/upgrade/incompatible.md`

tusaryan committed:
[Fix-17157][Master] Support setting max.concurrent.workflow.instances

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java:
```diff
@@ -18,10 +18,48 @@
 
+    private Integer maxConcurrentWorkflowInstances;
+
+    private WorkflowCacheRepository workflowCacheRepository;
```
```suggestion
    private IWorkflowRepository workflowRepository;
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java:
```diff
@@ -18,10 +18,48 @@
+
+    private WorkflowCacheRepository workflowCacheRepository;
+
+    public void setWorkflowCacheRepository(WorkflowCacheRepository workflowCacheRepository) {
+        this.workflowCacheRepository = workflowCacheRepository;
+    }
```
```suggestion
    
```

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java:
```diff
@@ -18,10 +18,48 @@
+        }
+
+        // Check workflow instance count if configured and repository is available
+        if (maxConcurrentWorkflowInstances != null && workflowCacheRepository != null) {
```
How about set a default value for `maxConcurrentWorkflowInstances`, e.g. `Integer.MAX_VALUE` then we don't need to check null here.

ruanwenjun reviewed (commented):
LGTM, and it's better to expose the `activeWorkflowInstanceCount` in `MasterHeartBeat`.

tusaryan committed:
Update MasterServerLoadProtection.java

tusaryan committed:
Update MasterServerLoadProtectionTest.java

tusaryan committed:
Update MasterServerLoadProtectionConfig.java

tusaryan commented:
Can anyone review the changes again

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java:
```diff
@@ -18,10 +18,44 @@
+        }
+
+        // Check workflow instance count if repository is available
+        if (workflowRepository != null) {
```
```suggestion

```
We don't need to check here, this component is must.

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionConfig.java:
```diff
@@ -0,0 +1,45 @@
+    @PostConstruct
+    public void init() {
+        MasterServerLoadProtection serverLoadProtection = masterConfig.getServerLoadProtection();
+        serverLoadProtection.setWorkflowRepository(workflowRepository);
```
It's better to inject at the constructor.

ruanwenjun replied:
Is this resolved? Why mark this as resolved?

tusaryan committed:
Update MasterServerLoadProtection.java

tusaryan committed:
Update MasterServerLoadProtectionConfig.java

yingh0ng commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtection.java:
```diff
@@ -18,10 +18,41 @@
+
+    @Override
+    public boolean isOverload(SystemMetrics systemMetrics) {
+        if (!enabled) {
```
Is this necessary? It duplicates in the `super.isOverLoad()` method.

ruanwenjun replied:
This is needed, since subclass has its own check rule.

ruanwenjun commented:
Please solve the code check style by `mvn spotless:apply`

tusaryan commented:
I wanted to let you know that I'm currently traveling. Because of this, I might be a bit delayed in addressing the feedback on this pull request. I will definitely get to it as soon as things settle down for me.

tusaryan committed:
Update MasterServerLoadProtection.java

tusaryan committed:
Update MasterServerLoadProtectionConfig.java

apply mvn spotless:apply

tusaryan commented:
> Please solve the code check style by `mvn spotless:apply`

Code style issues have been fixed by running mvn spotless:apply. Please let me know if there's anything else I can do!

tusaryan committed:
fix typo in MasterServerLoadProtection

tusaryan committed:
Inject MasterServerLoadProtection dependencies via constructor

- Refactored MasterServerLoadProtection to require IWorkflowRepository and maxConcurrentWorkflowInstances via constructor injection.
- Updated MasterServerLoadProtectionConfig to provide these dependencies using Spring @Bean and @Value with a default of 2147483647.
- Removed direct instantiation of MasterServerLoadProtection in MasterConfig to avoid missing constructor arguments.

tusaryan committed:
Apply spotless formatting

ruanwenjun reviewed (dismissed):
LGTM

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/config/MasterServerLoadProtectionConfig.java:
```diff
@@ -0,0 +1,43 @@
+    @Bean
+    public MasterServerLoadProtection masterServerLoadProtection(
+                                                                 IWorkflowRepository workflowRepository,
+                                                                 @Value("${master.server-load-protection.max-concurrent-workflow-instances:2147483647}") int maxConcurrentWorkflowInstances) {
```
We need to add this config in doc, and you need to set other fields in MasterServerLoadProtection, e.g. `maxSystemCpuUsagePercentageThresholds`, `maxJvmCpuUsagePercentageThresholds`.

tusaryan replied:
> We need to add this config in doc, and you need to set other fields in MasterServerLoadProtection, e.g. `maxSystemCpuUsagePercentageThresholds`, `maxJvmCpuUsagePercentageThresholds`.

Thanks for the feedback! I’ll get to it as soon as possible.

tusaryan replied:
Sorry for the delay. I will update it within 1-2 days.

tusaryan committed:
Add system metrics threshold configs and update docs for MasterServerLoadProtection

tusaryan commented:
I've integrated system metrics into MasterServerLoadProtection and updated its configuration documentation. Could someone please review the changes?

SbloodyS reviewed (dismissed):
+1

SbloodyS commented:
Please check the failed CI @tusaryan

tusaryan commented:
> Please check the failed CI @tusaryan

Okay, I'll check it out.

tusaryan committed:
Fix[Master]: Correct NullPointerException in MasterServerLoadProtectionTest

Stub IWorkflowRepository mock in tests to prevent NullPointerException when checking server load.

tusaryan committed:
Fix[Master]: resolve CI errors by updating test configuration for MasterServerLoadProtection

- Updated Spring test configuration in MasterConfigTest to properly autowire MasterServerLoadProtection and its dependencies.
- Added mock bean for IWorkflowRepository in tests.
- Provided test property overrides for load protection thresholds.

tusaryan committed:
apply Spotless for code style and formatting

tusaryan commented:
I've reviewed the logs, and the "Backend-Build" error appears unrelated to my changes. Could someone please help me with this?

ruanwenjun reviewed (approved):
LGTM

SbloodyS reviewed (approved):
+1
